src/cascadia/TerminalSettingsModel/VsSetupConfiguration.h

```diff
@@ -0,0 +1,218 @@
+            {
+                // The path of Microsoft.VisualStudio.DevShell.dll changed in 16.3
+                if (VersionInRange(L"[16.3,"))
+                {
+                    return ResolvePath(L"Common7\\Tools\\Microsoft.VisualStudio.DevShell.dll");
+                }
+
+                return ResolvePath(L"Common7\\Tools\\vsdevshell\\Microsoft.VisualStudio.DevShell.dll");
+            }
+
+            inline std::wstring GetDevCmdScriptPath() const
+            {
+                return ResolvePath(L"Common7\\Tools\\VsDevCmd.bat");
+            }
+
+            inline bool VersionInRange(std::wstring_view range) const
+            {
+                return VsSetupConfiguration::InstallationVersionInRange(query, inst, range);
+            }
+
+            inline std::wstring GetVersion() const
+            {
+                return VsSetupConfiguration::GetInstallationVersion(inst);
+            }
+
+            inline std::wstring GetInstallationPath() const
+            {
+                return VsSetupConfiguration::GetInstallationPath(inst);
+            }
+
+            inline std::wstring GetInstanceId() const
+            {
+                return VsSetupConfiguration::GetInstanceId(inst);
+            }
+
+            inline ComPtrPropertyStore GetInstancePropertyStore() const
+            {
+                return inst.query<ISetupPropertyStore>();
+            }
+
+            inline ComPtrCustomPropertyStore GetCustomPropertyStore() const
+            {
+                ComPtrCustomPropertyStore properties;
+                ComPtrSetupInstance2 instance2 = inst.query<ISetupInstance2>();
+                THROW_IF_FAILED(instance2->GetProperties(&properties));
+                return properties;
+            }
+
+            inline ComPtrCatalogPropertyStore GetCatalogPropertyStore() const
+            {
+                ComPtrCatalogPropertyStore properties;
+                ComPtrInstanceCatalog instanceCatalog = inst.query<ISetupInstanceCatalog>();
+                THROW_IF_FAILED(instanceCatalog->GetCatalogInfo(&properties));
+                return properties;
+            }
+
+            inline std::wstring GetProfileNameSuffix() const
+            {
+                return profileNameSuffix;
+            }
+
+        private:
+            friend class VsSetupConfiguration;
+
+            VsSetupInstance(const ComPtrSetupQuery pQuery, const ComPtrSetupInstance pInstance) :
+                query(pQuery),
+                helper(pQuery.query<ISetupHelper>()),
+                inst(pInstance),
+                profileNameSuffix(BuildProfileNameSuffix())
+            {
+            }
+
+            const ComPtrSetupQuery query;
+            const ComPtrSetupHelper helper;
+            const ComPtrSetupInstance inst;
+
+            std::wstring profileNameSuffix;
+
+            inline std::wstring BuildProfileNameSuffix() const
+            {
+                if (VersionInRange(L"[15.3,"))

```

Comment:
Now that I get down here, I see this *should* at least avoid the exceptions since an instance of this version would imply (barring anything strange a user did to their machine w.r.t. the COM server) the newer version of the COM server is installed.

However, if later a newer version of VS is installed side-by-side and the 15.0 - 15.2 is serviced by the newer engine, catalog files and support for reading them will be added so this code *could* work.

At this point, which this condition, I see nothing blocking this PR now (from a VS Setup viewpoint) but I would recommend considering relying on only the detection of `ISetupInstanceCatalog` whether you go down into this code path. It would be cleaner and more direct, not to mention a little faster since you're not having to parse and compare every instance version.

Reply:
Reduced exceptions, passing back nullptr and empty strings instead.